File: backend/src/probe.cpp
#include "probe.hpp"
#include "parser.hpp"

#include <cstdio>
#include <iostream>

#include <gst/gst.h>
#include <gst/mpegts/mpegts.h>
#include <glib.h>

using namespace Ats;

Probe::Probe(int s, string uri) : stream(s), uri(uri), metadata(s) {

    auto src   = Gst::ElementFactory::create_element("udpsrc");
    auto parse = Gst::ElementFactory::create_element("tsparse");
    auto sink  = Gst::ElementFactory::create_element("fakesink");

    pipe   = Gst::Pipeline::create();

    pipe->add(src)->add(parse)->add(sink);

    src->link(parse)->link(sink);

    src->set_property("uri", uri);
    src->set_property("timeout", 5000000000);

    bus    = pipe->get_bus();

    bus->add_watch(sigc::mem_fun(this, &Probe::on_bus_message));
}

void
Probe::set_state(Gst::State s) {
    if (pipe) {
        pipe->set_state(s);
    } else {
        throw No_pipe();
    }
}

bool
Probe::on_bus_message(const Glib::RefPtr<Gst::Bus>& bus,
                      const Glib::RefPtr<Gst::Message>& msg) {
    GstMpegtsSection*   section;
  
    switch (msg->get_message_type()) {
    case Gst::MESSAGE_ELEMENT: {
        if ((msg->get_source()->get_name().substr(0, 11) == "mpegtsparse") &&
            (section = gst_message_parse_mpegts_section (msg->gobj()))) {

            if(Parse::table(section, metadata))
                updated.emit(metadata);
	    
            gst_mpegts_section_unref (section);
        } else {
            if (msg->get_structure().get_name() == "GstUDPSrcTimeout") {

                if (! metadata.is_empty()) {
                    metadata.clear();
                    pipe->set_state(Gst::STATE_NULL);
                    pipe->set_state(Gst::STATE_PLAYING);
                    updated.emit(metadata);
                }
            }
        }
    }
        break;
    default: break;
    }
    return    true;
}

File: backend/src/probe.rs
use std::sync::{Arc,Mutex};
use signals::Signal;
use metadata::Structure;
use gst;
use gst::prelude::*;
use glib;
//use gst::DebugGraphDetails;
use gst_mpegts_sys;
use parse;
use initial::{Id,Uri};

pub struct Probe {
    pub stream:   String,
    pub updated:  Arc<Mutex<Signal<Structure>>>,

    pipeline:     gst::Pipeline,
}

impl Probe {
    pub fn new (stream : &(Id, Uri)) -> Probe {        
        let updated       = Arc::new(Mutex::new(Signal::new()));
        let id            = stream.0.clone();
        let uri           = stream.1.clone();
        let mut structure = Structure::new(uri.clone(), id.clone());
        
        let src   = gst::ElementFactory::make("udpsrc", None).unwrap();
        let parse = gst::ElementFactory::make("tsparse", None).unwrap();
        let sink  = gst::ElementFactory::make("fakesink", None).unwrap();

        let pipeline = gst::Pipeline::new(None);

        src.set_property("uri", &uri).unwrap();
        src.set_property("timeout", &5_000_000_000u64).unwrap();
        
        pipeline.add_many(&[&src, &parse, &sink]).unwrap();

        gst::Element::link_many(&[&src, &parse, &sink]).unwrap();
        
        let bus = pipeline.get_bus().unwrap();

        //let metadata  = structure.clone();
        let signal    = updated.clone();
        let pipe      = pipeline.clone();
        bus.add_watch(move |_, msg| {
            use gst::MessageView;
            
            match msg.view() {
                MessageView::Element(_) =>
                    if msg.get_src().unwrap().get_name().starts_with("mpegtsparse") {
                        unsafe {
                            let section = gst_mpegts_sys::gst_message_parse_mpegts_section(msg.as_mut_ptr());
                            if let Some(s) = parse::table(section, &mut structure) {
                                signal.lock().unwrap().emit(&s)
                            };
                        }
                    } else if let Some(s) = msg.get_structure() {
                            if s.get_name() == "GstUDPSrcTimeout" &&
                                ! structure.is_empty() {
                                    structure.clear();
                                    let _ = pipe.set_state(gst::State::Null);
                                    let _ = pipe.set_state(gst::State::Playing);
                                    signal.lock().unwrap().emit(&structure)
                                }
                    },
                _ => ()
            };
            glib::Continue(true)
        });
        
        Probe { stream : id, updated, pipeline }
    }

    pub fn set_state (&mut self, st: gst::State) {
        let _ = self.pipeline.set_state(st);
        //gst::debug_bin_to_dot_file(&self.pipeline, gst::DebugGraphDetails::VERBOSE, "pipeline");
    }
}
